From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
Subject: [PATCH] Remove dependency on `paste`

`paste` is unmaintained and was polluting `cargo audit` for Zephyr. Usage is
relatively minimal, so let's just remove it.

---

diff --git a/src/cppgc.rs b/src/cppgc.rs
index df2f27b..44fbf7b 100644
--- a/src/cppgc.rs
+++ b/src/cppgc.rs
@@ -452,116 +452,123 @@ impl<T: GarbageCollected> GetRustObj<T> for *mut RustObj {
 }
 
 macro_rules! member {
-  ($( # $attr:tt )* $name:ident) => {
-    paste::paste! {
-      #[repr(transparent)]
-      struct [< $name Inner >]([u8; crate::binding:: [< cppgc__ $name _SIZE >]]);
-
-      impl [< $name Inner >] {
-        fn new(ptr: *mut RustObj) -> Self {
-          let mut this = std::mem::MaybeUninit::uninit();
-          unsafe {
-            [< cppgc__ $name __CONSTRUCT >](this.as_mut_ptr(), ptr);
-            this.assume_init()
-          }
+  (
+    $( # $attr:tt )* $name:ident;
+    Inner = $inner:ident,
+    SIZE = $size:ident,
+    construct = $construct:ident,
+    get = $get:ident,
+    assign = $assign:ident,
+    destruct = $destruct:ident,
+    trace = $trace:ident
+  ) => {
+    #[repr(transparent)]
+    struct $inner([u8; crate::binding::$size]);
+
+    impl $inner {
+      fn new(ptr: *mut RustObj) -> Self {
+        let mut this = std::mem::MaybeUninit::uninit();
+        unsafe {
+          $construct(this.as_mut_ptr(), ptr);
+          this.assume_init()
         }
+      }
 
-        #[inline(always)]
-        fn get(&self) -> *mut RustObj {
-          // Member may be a compressed pointer, so just read it from C++
-          unsafe { [< cppgc__ $name __Get >](self) }
-        }
+      #[inline(always)]
+      fn get(&self) -> *mut RustObj {
+        // Member may be a compressed pointer, so just read it from C++
+        unsafe { $get(self) }
+      }
 
-        #[inline(always)]
-        fn assign(&mut self, ptr: *mut RustObj) {
-          // Assignment has write barriers in the GC, so call into C++
-          unsafe {
-            [< cppgc__ $name __Assign >](self, ptr);
-          }
+      #[inline(always)]
+      fn assign(&mut self, ptr: *mut RustObj) {
+        // Assignment has write barriers in the GC, so call into C++
+        unsafe {
+          $assign(self, ptr);
         }
       }
+    }
 
-      impl Drop for [< $name Inner >] {
-        fn drop(&mut self) {
-          unsafe {
-            [< cppgc__ $name __DESTRUCT >](self);
-          }
+    impl Drop for $inner {
+      fn drop(&mut self) {
+        unsafe {
+          $destruct(self);
         }
       }
+    }
 
-      $( # $attr )*
-      #[repr(transparent)]
-      pub struct $name<T: GarbageCollected> {
-        inner: [< $name Inner >],
-        _phantom: PhantomData<T>,
-      }
+    $( # $attr )*
+    #[repr(transparent)]
+    pub struct $name<T: GarbageCollected> {
+      inner: $inner,
+      _phantom: PhantomData<T>,
+    }
 
-      impl<T: GarbageCollected> $name<T> {
-        #[doc = "Create a new empty "]
-        #[doc = stringify!($name)]
-        #[doc = " which may be set later."]
-        pub fn empty() -> Self {
-          Self {
-            inner: [< $name Inner >]::new(std::ptr::null_mut()),
-            _phantom: PhantomData,
-          }
+    impl<T: GarbageCollected> $name<T> {
+      #[doc = "Create a new empty "]
+      #[doc = stringify!($name)]
+      #[doc = " which may be set later."]
+      pub fn empty() -> Self {
+        Self {
+          inner: $inner::new(std::ptr::null_mut()),
+          _phantom: PhantomData,
         }
+      }
 
-        #[doc = "Create a new "]
-        #[doc = stringify!($name)]
-        #[doc = " and initialize it with an object."]
-        pub fn new(other: &impl GetRustObj<T>) -> Self {
-          Self {
-            inner: [< $name Inner >]::new(other.get_rust_obj()),
-            _phantom: PhantomData,
-          }
+      #[doc = "Create a new "]
+      #[doc = stringify!($name)]
+      #[doc = " and initialize it with an object."]
+      pub fn new(other: &impl GetRustObj<T>) -> Self {
+        Self {
+          inner: $inner::new(other.get_rust_obj()),
+          _phantom: PhantomData,
         }
+      }
 
-        #[doc = "Set the object pointed to by this "]
-        #[doc = stringify!($name)]
-        #[doc = "."]
-        pub fn set(&mut self, other: &impl GetRustObj<T>) {
-          let ptr = other.get_rust_obj();
-          self.inner.assign(ptr);
-        }
+      #[doc = "Set the object pointed to by this "]
+      #[doc = stringify!($name)]
+      #[doc = "."]
+      pub fn set(&mut self, other: &impl GetRustObj<T>) {
+        let ptr = other.get_rust_obj();
+        self.inner.assign(ptr);
+      }
 
-        #[doc = "Get the object pointed to by this "]
-        #[doc = stringify!($name)]
-        #[doc = ", returning `None` if the pointer is empty or has been garbage-collected."]
-        #[doc = ""]
-        #[doc = "# Safety"]
-        #[doc = ""]
-        #[doc = "The caller must ensure that this pointer is being traced correctly by appearing in the [`trace`](GarbageCollected::trace)"]
-        #[doc = "implementation of the object that owns the pointer. Between initializing the pointer and calling `get()`, the pointer must be reachable by the garbage collector."]
-        pub unsafe fn get(&self) -> Option<&T> {
-          let ptr = self.inner.get();
-          if ptr.is_null() {
-            None
-          } else {
-            // SAFETY: Either this is a strong reference and the pointer is valid according
-            // to the safety contract of this method, or this is a weak reference and the
-            // ptr will be null if it was collected.
-            Some(unsafe { &*get_value_from_rust_obj(ptr) })
-          }
+      #[doc = "Get the object pointed to by this "]
+      #[doc = stringify!($name)]
+      #[doc = ", returning `None` if the pointer is empty or has been garbage-collected."]
+      #[doc = ""]
+      #[doc = "# Safety"]
+      #[doc = ""]
+      #[doc = "The caller must ensure that this pointer is being traced correctly by appearing in the [`trace`](GarbageCollected::trace)"]
+      #[doc = "implementation of the object that owns the pointer. Between initializing the pointer and calling `get()`, the pointer must be reachable by the garbage collector."]
+      pub unsafe fn get(&self) -> Option<&T> {
+        let ptr = self.inner.get();
+        if ptr.is_null() {
+          None
+        } else {
+          // SAFETY: Either this is a strong reference and the pointer is valid according
+          // to the safety contract of this method, or this is a weak reference and the
+          // ptr will be null if it was collected.
+          Some(unsafe { &*get_value_from_rust_obj(ptr) })
         }
       }
+    }
 
-      impl<T: GarbageCollected> GetRustObj<T> for $name<T> {
-        fn get_rust_obj(&self) -> *mut RustObj {
-          self.inner.get()
-        }
+    impl<T: GarbageCollected> GetRustObj<T> for $name<T> {
+      fn get_rust_obj(&self) -> *mut RustObj {
+        self.inner.get()
       }
+    }
 
-      impl<T: GarbageCollected> Traced for $name<T> {
-        fn trace(&self, visitor: &mut Visitor) {
-          unsafe { [< cppgc__Visitor__Trace__ $name >](visitor, &self.inner) }
-        }
+    impl<T: GarbageCollected> Traced for $name<T> {
+      fn trace(&self, visitor: &mut Visitor) {
+        unsafe { $trace(visitor, &self.inner) }
       }
+    }
 
-      impl<T: GarbageCollected> std::fmt::Debug for $name<T> {
-        fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
-          fmt.debug_struct(stringify!($name)).finish()
-        }
+    impl<T: GarbageCollected> std::fmt::Debug for $name<T> {
+      fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        fmt.debug_struct(stringify!($name)).finish()
       }
     }
   }
@@ -571,7 +578,14 @@ member! {
   /// Members are used in classes to contain strong pointers to other garbage
   /// collected objects. All Member fields of a class must be traced in the class'
   /// trace method.
-  Member
+  Member;
+  Inner = MemberInner,
+  SIZE = cppgc__Member_SIZE,
+  construct = cppgc__Member__CONSTRUCT,
+  get = cppgc__Member__Get,
+  assign = cppgc__Member__Assign,
+  destruct = cppgc__Member__DESTRUCT,
+  trace = cppgc__Visitor__Trace__Member
 }
 
 member! {
@@ -581,95 +595,107 @@ member! {
   /// pointee alive. Hence if all pointers to to a heap allocated object are weak
   /// the object will be garbage collected. At the time of GC the weak pointers
   /// will automatically be set to null.
-  WeakMember
+  WeakMember;
+  Inner = WeakMemberInner,
+  SIZE = cppgc__WeakMember_SIZE,
+  construct = cppgc__WeakMember__CONSTRUCT,
+  get = cppgc__WeakMember__Get,
+  assign = cppgc__WeakMember__Assign,
+  destruct = cppgc__WeakMember__DESTRUCT,
+  trace = cppgc__Visitor__Trace__WeakMember
 }
 
 macro_rules! persistent {
-  ($( # $attr:tt )* $name:ident) => {
-    paste::paste! {
-      // PersistentBase is extremely particular about move and copy semantics,
-      // so we allocate it on the heap and only interact with it via calls to C++.
-      #[repr(C)]
-      struct [< $name Inner >](Opaque);
-
-      $( # $attr )*
-      pub struct $name<T: GarbageCollected> {
-        inner: *mut [< $name Inner >],
-        _phantom: PhantomData<T>,
-      }
+  (
+    $( # $attr:tt )* $name:ident;
+    Inner = $inner:ident,
+    construct = $construct:ident,
+    assign = $assign:ident,
+    get = $get:ident,
+    destruct = $destruct:ident
+  ) => {
+    // PersistentBase is extremely particular about move and copy semantics,
+    // so we allocate it on the heap and only interact with it via calls to C++.
+    #[repr(C)]
+    struct $inner(Opaque);
+
+    $( # $attr )*
+    pub struct $name<T: GarbageCollected> {
+      inner: *mut $inner,
+      _phantom: PhantomData<T>,
+    }
 
-      impl<T: GarbageCollected> $name<T> {
-        #[doc = "Create a new empty "]
-        #[doc = stringify!($name)]
-        #[doc = " which may be set later."]
-        pub fn empty() -> Self {
-          let this = unsafe { [< cppgc__ $name __CONSTRUCT >](std::ptr::null_mut()) };
-          Self {
-            inner: this,
-            _phantom: PhantomData,
-          }
+    impl<T: GarbageCollected> $name<T> {
+      #[doc = "Create a new empty "]
+      #[doc = stringify!($name)]
+      #[doc = " which may be set later."]
+      pub fn empty() -> Self {
+        let this = unsafe { $construct(std::ptr::null_mut()) };
+        Self {
+          inner: this,
+          _phantom: PhantomData,
         }
+      }
 
-        #[doc = "Create a new "]
-        #[doc = stringify!($name)]
-        #[doc = " and initialize it with an object."]
-        pub fn new(other: &impl GetRustObj<T>) -> Self {
-          let this = unsafe { [< cppgc__ $name __CONSTRUCT >](other.get_rust_obj()) };
-          Self {
-            inner: this,
-            _phantom: PhantomData,
-          }
+      #[doc = "Create a new "]
+      #[doc = stringify!($name)]
+      #[doc = " and initialize it with an object."]
+      pub fn new(other: &impl GetRustObj<T>) -> Self {
+        let this = unsafe { $construct(other.get_rust_obj()) };
+        Self {
+          inner: this,
+          _phantom: PhantomData,
         }
+      }
 
-        #[doc = "Set the object pointed to by this "]
-        #[doc = stringify!($name)]
-        #[doc = "."]
-        pub fn set(&mut self, other: &impl GetRustObj<T>) {
-          let ptr = other.get_rust_obj();
-          self.assign(ptr);
-        }
+      #[doc = "Set the object pointed to by this "]
+      #[doc = stringify!($name)]
+      #[doc = "."]
+      pub fn set(&mut self, other: &impl GetRustObj<T>) {
+        let ptr = other.get_rust_obj();
+        self.assign(ptr);
+      }
 
-        #[doc = "Borrow the object pointed to by this "]
-        #[doc = stringify!($name)]
-        #[doc = "."]
-        pub fn get(&self) -> Option<&T> {
-          let ptr = self.get_rust_obj();
-          if ptr.is_null() {
-            None
-          } else {
-            // SAFETY: Either this is a strong reference and the pointer is always valid
-            // or this is a weak reference and the ptr will be null if it was collected.
-            Some(unsafe { &*get_value_from_rust_obj(ptr) })
-          }
+      #[doc = "Borrow the object pointed to by this "]
+      #[doc = stringify!($name)]
+      #[doc = "."]
+      pub fn get(&self) -> Option<&T> {
+        let ptr = self.get_rust_obj();
+        if ptr.is_null() {
+          None
+        } else {
+          // SAFETY: Either this is a strong reference and the pointer is always valid
+          // or this is a weak reference and the ptr will be null if it was collected.
+          Some(unsafe { &*get_value_from_rust_obj(ptr) })
         }
+      }
 
-        #[inline(always)]
-        fn assign(&mut self, ptr: *mut RustObj) {
-          unsafe {
-            [< cppgc__ $name __Assign >](self.inner, ptr);
-          }
+      #[inline(always)]
+      fn assign(&mut self, ptr: *mut RustObj) {
+        unsafe {
+          $assign(self.inner, ptr);
         }
       }
+    }
 
-      impl<T: GarbageCollected> Drop for $name<T> {
-        fn drop(&mut self) {
-          unsafe {
-            [< cppgc__ $name __DESTRUCT >](self.inner);
-          }
+    impl<T: GarbageCollected> Drop for $name<T> {
+      fn drop(&mut self) {
+        unsafe {
+          $destruct(self.inner);
         }
       }
+    }
 
-      impl<T: GarbageCollected> std::fmt::Debug for $name<T> {
-        fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
-          fmt.debug_struct(stringify!($name)).finish()
-        }
+    impl<T: GarbageCollected> std::fmt::Debug for $name<T> {
+      fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        fmt.debug_struct(stringify!($name)).finish()
       }
+    }
 
-      impl<T: GarbageCollected> GetRustObj<T> for $name<T> {
-        fn get_rust_obj(&self) -> *mut RustObj {
-          unsafe {
-            [< cppgc__ $name __Get >](self.inner)
-          }
+    impl<T: GarbageCollected> GetRustObj<T> for $name<T> {
+      fn get_rust_obj(&self) -> *mut RustObj {
+        unsafe {
+          $get(self.inner)
         }
       }
     }
@@ -682,7 +708,12 @@ persistent! {
   /// keep the object pointed to alive. The Persistent handle is always a GC root
   /// from the point of view of the GC. Persistent must be constructed and
   /// destructed in the same thread.
-  Persistent
+  Persistent;
+  Inner = PersistentInner,
+  construct = cppgc__Persistent__CONSTRUCT,
+  assign = cppgc__Persistent__Assign,
+  get = cppgc__Persistent__Get,
+  destruct = cppgc__Persistent__DESTRUCT
 }
 
 persistent! {
@@ -690,7 +721,12 @@ persistent! {
   /// an on-heap object. The pointer is automatically cleared when the pointee gets
   /// collected. WeakPersistent must be constructed and destructed in the same
   /// thread.
-  WeakPersistent
+  WeakPersistent;
+  Inner = WeakPersistentInner,
+  construct = cppgc__WeakPersistent__CONSTRUCT,
+  assign = cppgc__WeakPersistent__Assign,
+  get = cppgc__WeakPersistent__Get,
+  destruct = cppgc__WeakPersistent__DESTRUCT
 }
 
 /// `GcCell` is a zero-overhead memory cell that provides interior mutability
diff --git a/src/typed_array.rs b/src/typed_array.rs
index 497359c..781f335 100644
--- a/src/typed_array.rs
+++ b/src/typed_array.rs
@@ -5,7 +5,6 @@ use crate::PinScope;
 use crate::TypedArray;
 use crate::binding::v8__TypedArray__kMaxByteLength;
 use crate::support::size_t;
-use paste::paste;
 
 unsafe extern "C" {
   fn v8__TypedArray__Length(this: *const TypedArray) -> size_t;
@@ -25,45 +24,43 @@ impl TypedArray {
 }
 
 macro_rules! typed_array {
-  ($name:ident) => {
-    paste! {
-      use crate::$name;
+  ($name:ident; new = $new:ident, max_length = $max_length:ident) => {
+    use crate::$name;
 
-      unsafe extern "C" {
-        fn [< v8__ $name __New >](
-          buf_ptr: *const ArrayBuffer,
-          byte_offset: usize,
-          length: usize,
-        ) -> *const $name;
-      }
-
-      impl $name {
-        #[inline(always)]
-        pub fn new<'s>(
-          scope: &PinScope<'s, '_>,
-          buf: Local<ArrayBuffer>,
-          byte_offset: usize,
-          length: usize,
-        ) -> Option<Local<'s, $name>> {
-          unsafe { scope.cast_local(|_| [< v8__ $name __New >](&*buf, byte_offset, length)) }
-        }
+    unsafe extern "C" {
+      fn $new(
+        buf_ptr: *const ArrayBuffer,
+        byte_offset: usize,
+        length: usize,
+      ) -> *const $name;
+    }
 
-        #[doc = concat!("The largest ", stringify!($name), " size that can be constructed using `new`.")]
-        pub const MAX_LENGTH: usize = crate::binding::[< v8__ $name __kMaxLength >];
+    impl $name {
+      #[inline(always)]
+      pub fn new<'s>(
+        scope: &PinScope<'s, '_>,
+        buf: Local<ArrayBuffer>,
+        byte_offset: usize,
+        length: usize,
+      ) -> Option<Local<'s, $name>> {
+        unsafe { scope.cast_local(|_| $new(&*buf, byte_offset, length)) }
       }
+
+      #[doc = concat!("The largest ", stringify!($name), " size that can be constructed using `new`.")]
+      pub const MAX_LENGTH: usize = crate::binding::$max_length;
     }
   };
 }
 
-typed_array!(Uint8Array);
-typed_array!(Uint8ClampedArray);
-typed_array!(Int8Array);
-typed_array!(Uint16Array);
-typed_array!(Int16Array);
-typed_array!(Uint32Array);
-typed_array!(Int32Array);
-typed_array!(Float16Array);
-typed_array!(Float32Array);
-typed_array!(Float64Array);
-typed_array!(BigUint64Array);
-typed_array!(BigInt64Array);
+typed_array!(Uint8Array; new = v8__Uint8Array__New, max_length = v8__Uint8Array__kMaxLength);
+typed_array!(Uint8ClampedArray; new = v8__Uint8ClampedArray__New, max_length = v8__Uint8ClampedArray__kMaxLength);
+typed_array!(Int8Array; new = v8__Int8Array__New, max_length = v8__Int8Array__kMaxLength);
+typed_array!(Uint16Array; new = v8__Uint16Array__New, max_length = v8__Uint16Array__kMaxLength);
+typed_array!(Int16Array; new = v8__Int16Array__New, max_length = v8__Int16Array__kMaxLength);
+typed_array!(Uint32Array; new = v8__Uint32Array__New, max_length = v8__Uint32Array__kMaxLength);
+typed_array!(Int32Array; new = v8__Int32Array__New, max_length = v8__Int32Array__kMaxLength);
+typed_array!(Float16Array; new = v8__Float16Array__New, max_length = v8__Float16Array__kMaxLength);
+typed_array!(Float32Array; new = v8__Float32Array__New, max_length = v8__Float32Array__kMaxLength);
+typed_array!(Float64Array; new = v8__Float64Array__New, max_length = v8__Float64Array__kMaxLength);
+typed_array!(BigUint64Array; new = v8__BigUint64Array__New, max_length = v8__BigUint64Array__kMaxLength);
+typed_array!(BigInt64Array; new = v8__BigInt64Array__New, max_length = v8__BigInt64Array__kMaxLength);
